src/lazy.jl

```diff
@@ -440,7 +441,18 @@ function Base.convert(::Type{T}, x::PtrString) where {T <: Enum}
     throw(ArgumentError("invalid `$T` string value: \"$sym\""))
 end
 
-Base.:(==)(x::PtrString, y::AbstractString) = x.len == sizeof(y) && ccall(:memcmp, Cint, (Ptr{UInt8}, Ptr{UInt8}, Csize_t), x.ptr, pointer(y), x.len) == 0
+function Base.:(==)(x::PtrString, y::AbstractString)
+    # If the string is backed by UInt8 codeunits, the `sizeof` computation below is correct
+    return if codeunit(y) == UInt8
+        x.len == sizeof(y) || return false
+        ref = Base.cconvert(Ptr{UInt8}, y)
+        cmp = GC.@preserve ref ccall(:memcmp, Cint, (Ptr{UInt8}, Ptr{UInt8}, Csize_t), x.ptr, Base.unsafe_convert(Ptr{UInt8}, ref), x.len)
```

Comment:
The manual preserve here seems to be exactly equivalent to simply passing y to ccall directly.